lib/trivial.nix

```diff
@@ -232,6 +232,31 @@ in {
 
   nixpkgsVersion = builtins.trace "`lib.nixpkgsVersion` is deprecated, use `lib.version` instead!" version;
 
+  /* The store path to this nixpkgs as a string with context. This can be used
+     to incur a dependency on this nixpkgs, for example, to symlink it into a
+     NixOS closure.
+
+     This is different than pkgs.path if nixpkgs is already in the Nix store,
+     since pkgs.path is a path to nixpkgs rather than a string-with-context,
+     and thus its usage typically, undesirably, results in copying this
+     nixpkgs.
+
+     See: https://github.com/NixOS/nix/issues/5868
+          https://github.com/NixOS/nix/issues/9428
+  */
+  nixpkgsStorePathString = if lib.trivial.inPureEvalMode || !(lib.strings.isStorePath (toString ../.))
+    # If we are in pure eval mode and we are not a flake, we cannot generate
+    # the string with context. Not much we can do.
+    then "${../.}"
+    else builtins.storePath ../.;
+
+  /* Whether this nixpkgs is being used natively as a flake.
+
+     Note that this is not set for `import nixpkgs` inside a flake,
+     only for native flake usage.
```

Comment:
What's "native flake usage"? I guess this really just means using `legacyPackages` or the `lib` flake directly. I think this should be worded more clearly here.

Reply:
Yup. Changed.